client/src/slate-helpers/slate-change-mutations/scanBlockAndConvertOuterSquareBrackets.js

```diff
@@ -0,0 +1,118 @@
+export function scanBlockAndConvertOuterSquareBrackets ({ change, updateBlock, exportSelection, blockId }) {
+  const value = change.value;
+  const textNodes = value.document.getTexts();
+
+  const {
+    balancedBrackets,
+    startKey,
+    startOffset,
+    endKey,
+    endOffset,
+  } = findOutermostSquareBrackets(textNodes);
+
+  if (balancedBrackets) {
+    exportBracketedSelectionFromRange({
+      change,
+      updateBlock,
+      exportSelection,
+      blockId,
+      startKey,
+      startOffset,
+      endKey,
+      endOffset,
+     });
+    return true;
+  }
+  return false;
+}
+
+function exportBracketedSelectionFromRange ({
+  change,
+  updateBlock,
+  exportSelection,
+  blockId,
+  startKey,
+  startOffset,
+  endKey,
+  endOffset,
+}) {
+  // delete opening square bracket
+  change.select({
+    anchorKey: startKey,
+    anchorOffset: startOffset,
+    focusKey: startKey,
+    focusOffset: startOffset,
+  })
+  .deleteForward(1);
+
+  // delete closing square square bracket
+  // calculating achorOffset and focusOffset is complicated
+  // by the fact that we need watch out for both brackets being
+  // in the same text node, in which case deleting the opening square
+  // bracket will have changed the offset of the closing square bracket
+  change.select({
+    anchorKey: endKey,
+    anchorOffset: startKey === endKey ? endOffset - 1 : endOffset,
+    focusKey: endKey,
+    focusOffset: startKey === endKey ? endOffset - 1 : endOffset,
+  })
+  .deleteForward(1);
+
+  // select everything that was within the square brackets
+  change.select({
+    anchorKey: startKey,
+    anchorOffset: startOffset,
+    focusKey: endKey,
+    focusOffset: startKey === endKey ? endOffset - 1 : endOffset,
+  });
+
+  updateBlock({ id: blockId, value: change.value, pointerChanged: false });
+  exportSelection(blockId);
+};
+
+function findOutermostSquareBrackets(textNodes) {
+  let startKey;
+  let startOffset;
+  let endKey;
+  let endOffset;
+
+  let matchFound = false;
+  let squareBracketBalance = 0;
+
```

Comment:
I imagine there are ways to clean up the for/while loop below, but don't know if it's worth the time. If you can figure out something clever then kudos. One nice thing is that at least this function is very contained.

Reply:
I'll just leave it as it is for now, but if you think of a concrete improvement I'd be happy to incorporate it.